cmake/add_tests.cmake

```diff
@@ -19,55 +19,62 @@
 # OpenSSL libraries as published by the OpenSSL project. See the file
 # COPYING for details.
 #
-#############################################################################
-
+# ############################################################################
 
 include(CMakeParseArguments)
 
-if (BUILD_TESTING)
+if(BUILD_TESTING)
   include(CheckPIESupported)
   check_pie_supported()
 endif()
 
-function (add_unit_test)
-
-  if (NOT BUILD_TESTING)
+function(add_unit_test)
+  if(NOT BUILD_TESTING)
     return()
   endif()
 
-  cmake_parse_arguments(ADD_UNIT_TEST "CRITERION;LIBTEST" "TARGET" "SOURCES;DEPENDS;INCLUDES" ${ARGN})
+  cmake_parse_arguments(ADD_UNIT_TEST "CRITERION;LIBTEST" "TARGET" "SOURCES;DEPENDS;INCLUDES;MOCKS" ${ARGN})
 
-  if (NOT ADD_UNIT_TEST_SOURCES)
+  if(NOT ADD_UNIT_TEST_SOURCES)
     set(ADD_UNIT_TEST_SOURCES "${ADD_UNIT_TEST_TARGET}.c")
   endif()
 
+  if(ADD_UNIT_TEST_MOCKS)
+    set(ADD_UNIT_TEST_MOCKS_LIB "${ADD_UNIT_TEST_TARGET}_mocks")
+    add_library(${ADD_UNIT_TEST_MOCKS_LIB} SHARED "${ADD_UNIT_TEST_MOCKS}")
```

Comment:
Can you explain please, how this works?

When a mock is defined, it looks like both the original and the mocked symbol will exist in 2 different shared libraries. Which one will be used and why?

Reply:
Unlikely in the case of static linking, seems for shared libraries
- the link order decides where a unit-referenced function will be searched for first
so, the key here is 
```
   target_link_libraries(${ADD_UNIT_TEST_MOCKS_LIB} syslog-ng)
  endif()
  ...
  target_link_libraries(${ADD_UNIT_TEST_TARGET} ${ADD_UNIT_TEST_DEPENDS} ${ADD_UNIT_TEST_MOCKS_LIB} syslog-ng)

```
- the mocked functions in the mocking shared library are looking for and using any referenced functions in itself first
that is exactly our intended behavior.

(Telling the truth, I did not play along with all the variants, I was satisfied with the result that worked both on Linux and macOS.)

Reply:
This whole mock separation is revoked currently